src/state_machine/coordinator/fire.rs

```diff
@@ -0,0 +1,1296 @@
+                prev_state == &State::Idle
+                    || prev_state == &State::DkgPublicGather
+                    || prev_state == &State::DkgEndGather
+            }
+            State::DkgPublicGather => {
+                prev_state == &State::DkgPublicDistribute || prev_state == &State::DkgPublicGather
+            }
+            State::DkgPrivateDistribute => prev_state == &State::DkgPublicGather,
+            State::DkgEndGather => prev_state == &State::DkgPrivateDistribute,
+            State::NonceRequest(is_taproot, merkle_root) => {
+                prev_state == &State::Idle
+                    || prev_state == &State::DkgEndGather
+                    || prev_state == &State::SigShareGather(*is_taproot, *merkle_root)
+            }
+            State::NonceGather(is_taproot, merkle_root) => {
+                prev_state == &State::NonceRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareRequest(is_taproot, merkle_root) => {
+                prev_state == &State::NonceGather(*is_taproot, *merkle_root)
+            }
+            State::SigShareGather(is_taproot, merkle_root) => {
+                prev_state == &State::SigShareRequest(*is_taproot, *merkle_root)
+                    || prev_state == &State::SigShareGather(*is_taproot, *merkle_root)
+            }
+        };
+        if accepted {
+            debug!("state change from {:?} to {:?}", prev_state, state);
+            Ok(())
+        } else {
+            Err(Error::BadStateChange(format!(
+                "{:?} to {:?}",
+                prev_state, state
+            )))
+        }
+    }
+}
+
+impl<Aggregator: AggregatorTrait> CoordinatorTrait for Coordinator<Aggregator> {
+    /// Create a new coordinator
+    fn new(config: Config) -> Self {
+        Self {
+            aggregator: Aggregator::new(config.num_keys, config.threshold),
+            config,
+            current_dkg_id: 0,
+            current_sign_id: 0,
+            current_sign_iter_id: 0,
+            dkg_public_shares: Default::default(),
+            party_polynomials: Default::default(),
+            public_nonces: Default::default(),
+            signature_shares: Default::default(),
+            aggregate_public_key: None,
+            signature: None,
+            schnorr_proof: None,
+            message: Default::default(),
+            dkg_wait_signer_ids: Default::default(),
+            nonce_recv_key_ids: Default::default(),
+            sign_wait_signer_ids: Default::default(),
+            sign_recv_key_ids: Default::default(),
+            state: State::Idle,
+            nonce_start: None,
+            sign_start: None,
+            malicious_signer_ids: Default::default(),
+        }
+    }
+
+    /// Retrieve the config
+    fn get_config(&self) -> Config {
+        self.config.clone()
+    }
+
+    /// Process inbound messages
+    fn process_inbound_messages(
```

Comment:
This function seems common between the types of coordinators. Might make sense to rip it out. Am sure there are a few other functions that are, but understand if you want to keep code fully seperated between the two coordinator types.

Reply:
I need to pull the common code out into the trait, but that will require another private trait that allows access to the common fields.  So def not this PR but one soon.